ext/geos_c_impl/geometry_collection.c

```diff
@@ -64,8 +84,23 @@ create_geometry_collection(VALUE module, int type, VALUE factory, VALUE array)
       break;
   }
   for (i = 0; i < len; ++i) {
-    geom = rgeo_convert_to_detached_geos_geometry(
-      rb_ary_entry(array, i), factory, cast_type, &klass, &state);
+    args.obj = rb_ary_entry(array, i);
+    args.factory = factory;
+    args.type = cast_type;
+    args.klasses = &klass;
+    args.state = &state;
+
+    geom = rb_protect(rgeo_convert_to_detached_geos_geometry_wrapper,
```

Comment:
Since we're raising now, this function and the `coord_seq_from_array` function had memory leaks because we allocated data before calling `rgeo_convert_*` in each of them. This wraps those calls in `rb_protect` so that we can handle the error, free the existing data and then re-raise.

I could go with `rb_rescue`, but `rb_protect` feels a little nicer to handle all of the error logic within the original caller.